File: 0259-3sum-smaller/0259-3sum-smaller.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumSmaller = function(nums, target) {

    if(nums.length < 3 ) return 0;

    nums.sort((a,b) => a-b);

    //i 가 가리키는 부분

    let result = 0;

    for(let i = 0 ; i < nums.length - 2 ; i++){
        // 
        let left = i + 1; 
        let right = nums.length - 1;
        while(left < right){
            let sum = nums[i] + nums[left] + nums[right];

            // target 보다 같거나 작으면
            if(sum < target){
                result += (right - left); // right - left 사이의 조합이 유효하므로 
                left++;
            } else {
                right--;
            }
        }     
    }
    
    return result;
};

File: 0259-3sum-smaller/0259-3sum-smaller.go
import "sort"

func threeSumSmaller(nums []int, target int) int {
    if len(nums) < 3 {
        return 0
    }

    // 정렬 필요
    sort.Ints(nums)
    
    count := 0
    
    for i := 0; i < len(nums)-2; i++ {
        left, right := i+1, len(nums)-1
        
        for left < right {
            sum := nums[i] + nums[left] + nums[right]

            if sum < target {
                // (nums[left], nums[right]) 사이 모든 조합이 유효
                count += right - left
                left++ // 더 큰 값 탐색
            } else {
                right-- // 더 작은 값 탐색
            }
        }
    }
    
    return count
}